astropy/units/core.py

```diff
@@ -2093,38 +2093,39 @@ def __call__(
             if isinstance(s, bytes):
                 s = s.decode("ascii")
 
-            try:
-                return f.parse(s)
-            except NotImplementedError:
-                raise
-            except Exception as e:
-                if parse_strict == "silent":
-                    pass
-                else:
-                    # Deliberately not issubclass here. Subclasses
-                    # should use their name.
-                    if f is not unit_format.Generic:
-                        format_clause = f.name + " "
-                    else:
-                        format_clause = ""
-                    msg = (
-                        f"'{s}' did not parse as {format_clause}unit: {str(e)} "
-                        "If this is meant to be a custom unit, "
-                        "define it with 'u.def_unit'. To have it "
-                        "recognized inside a file reader or other code, "
-                        "enable it with 'u.add_enabled_units'. "
-                        "For details, see "
-                        "https://docs.astropy.org/en/latest/units/combining_and_defining.html"
+            with warnings.catch_warnings():
```

Comment:
Since our filter is simple, I think we can speed things up a bit by passing in arguments here rather than setting up a `filterwarnings` (saves almost a factor 2 in my tests). And maybe even better if we do something like:
```
WARNING_FILTER = {
    "warn": "default",
    "raise": "error",
    "silent": "ignore",
}
try:
    with warnings.catchwarnings(action=WARNING_FILTER[parse_strict], category=UnitParserWarning):
        return f.parse(s)
    except KeyError:
        if parse_strict not in WARNING_FILTER:
            raise ValueError(
                        "'parse_strict' must be 'warn', 'raise' or 'silent'"
                    ) from None
        raise
    except NotImplementedError:
        raise
...
```

Reply:
Maybe we should bite the bullet and set the minimum version to python 3.11 - we're dropping support for it in the next release anyway... (#16886 was also annoying because of this...)

Reply:
See #16900 for an issue about dropping python 3.10...

Reply:
The suggestion we are discussing here is about an implementation detail that can affect the performance of the code, but not the API. Is the proposed API good enough that I should start working on the documentation or is there anything that still needs to be discussed about the proposed API?

Reply:
Personally I'm significantly more in favor of the "Alternative API" you discuss. Avoiding / removing custom meta classes is almost always an improvement. Adding the kwarg will be a bit backwards incompatible, but we are coming up on a major version update, so that seems fine to me.